packages/foundry/contracts/hooks/rebalancer/Rebalancer.sol

```diff
@@ -0,0 +1,633 @@
+        hookFlags.shouldCallAfterRemoveLiquidity = true;
+        hookFlags.shouldCallComputeDynamicSwapFee = true;
+        hookFlags.shouldCallAfterSwap = true;
+        return hookFlags;
+    }
+
+    /// @inheritdoc BaseHooks
+    function onBeforeAddLiquidity(
+        address router,
+        address,
+        AddLiquidityKind,
+        uint256[] memory,
+        uint256,
+        uint256[] memory,
+        bytes memory
+    ) public view override onlySelfRouter(router) returns (bool) {
+        // We only allow addLiquidity via the Router/Hook itself
+        return true;
+    }
+
+    /// @inheritdoc BaseHooks
+    function onAfterRemoveLiquidity(
+        address router,
+        address pool,
+        RemoveLiquidityKind,
+        uint256,
+        uint256[] memory,
+        uint256[] memory amountsOutRaw,
+        uint256[] memory,
+        bytes memory userData
+    ) public override onlySelfRouter(router) returns (bool, uint256[] memory hookAdjustedAmountsOutRaw) {
+        return (true, amountsOutRaw);
+    }
+
+    /// @inheritdoc IHooks
+    function onAfterSwap(
+        AfterSwapParams calldata params
+    ) public override onlyVault returns (bool success, uint256 hookAdjustedAmountCalculatedRaw) {
+        address pool = params.pool;
+        if (rebalanceData[pool].length == 0) {
+            revert RebalanceDataNotSet(pool);
+        }
+
+        (bool rebalanceRequired, uint256[] memory priceActionRatio) = isRebalanceRequired(pool);
+        if (rebalanceRequired) {
+            emit RebalanceStarted(pool);
+            rebalance(pool, priceActionRatio);
+        }
+
+        return (true, params.amountCalculatedRaw);
+    }
+
+    /// @inheritdoc IHooks
+    function onComputeDynamicSwapFeePercentage(
+        PoolSwapParams calldata params,
+        address pool,
+        uint256 staticSwapFeePercentage
+    ) public view override onlyVault returns (bool, uint256) {
+        // uint256 dynamicFee = IOracle(oracle).getFee(pool);
+
+        return (true, staticSwapFeePercentage);
+    }
+
+    /***************************************************************************
+                                  Setter Functions
+    ***************************************************************************/
+    function setWeightedPoolFactoryAddress(address newWeightedPoolFactory) external onlyOwner {
+        ///@dev I am a little confused that whether to check if the factory is disable or not would love
+        /// reason to check why cause I am thinking the pool can be created from a disabled factory
+        /// and still work fine? Let's keep it like this for now while I study about balancer governance
+        /// more
+        weightedPoolFactory = newWeightedPoolFactory;
+    }
+
+    function setRebalanceData(address pool, RebalanceData[] memory _rebalanceData) external {
+        PoolRoleAccounts memory roleAccounts = _vault.getPoolRoleAccounts(pool);
+
+        if (msg.sender != roleAccounts.poolCreator) {
```

Comment:
This is mixing concerns a bit. The purpose of the pool creator is to set pool creator fees. You're using it for a different purpose here - and it prevents you from using it with standard WeightedPools, which can't have a pool creator (though this can be circumvented in test code with mocks).

Reply:
Got it didn't know weighted pools had no creators I guess a custom factory would be the way to go as this part should only be done by the creator.

I was able to get the test passed with vm.mock's so yeah